hw07_file_copying/copy.go

```diff
@@ -10,6 +15,95 @@ var (
 )
 
 func Copy(fromPath, toPath string, offset, limit int64) error {
-	// Place your code here.
+	sourceInfo, err := os.Stat(fromPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("source file %s does not exist: %w", fromPath, err)
+		}
+		return fmt.Errorf("error accessing source file: %w", err)
+	}
+
+	if !sourceInfo.Mode().IsRegular() {
+		return ErrUnsupportedFile
+	}
+
+	fileSize := sourceInfo.Size()
+
+	if offset > fileSize {
+		return ErrOffsetExceedsFileSize
+	}
+
+	remainingBytes := fileSize - offset
+
+	bytesToCopy := remainingBytes
+	if limit > 0 && limit < remainingBytes {
+		bytesToCopy = limit
+	}
+
+	sourceFile, err := os.Open(fromPath)
+	if err != nil {
+		return fmt.Errorf("error opening source file: %w", err)
+	}
+
+	defer sourceFile.Close()
+
+	destFile, err := os.Create(toPath)
+	if err != nil {
+		return fmt.Errorf("error creating destination file: %w", err)
+	}
+	defer destFile.Close()
+
+	_, err = sourceFile.Seek(offset, io.SeekStart)
+	if err != nil {
+		return fmt.Errorf("error seeking source file: %w", err)
+	}
+
+	// simple progress indicator (no external packages)
+	const chunkSize = 1 // 32 * 1024  -  чтобы проверять работу прогресс-бара проще было, в оригинале я 32кб использовал
+	buf := make([]byte, chunkSize)
+	var copied int64
+
+	for copied < bytesToCopy {
+		remaining := bytesToCopy - copied
+		if remaining < int64(len(buf)) {
+			buf = buf[:remaining]
+		}
+
+		n, readErr := sourceFile.Read(buf)
+		if n > 0 {
+			w, writeErr := destFile.Write(buf[:n])
+			if writeErr != nil {
+				return fmt.Errorf("error writing to destination file: %w", writeErr)
+			}
+			copied += int64(w)
+
+			// update progress bar
+			printProgressBar(copied, bytesToCopy, 40)
+		}
+
+		if readErr != nil {
+			if readErr == io.EOF {
+				break
+			}
+			return fmt.Errorf("error reading from source file: %w", readErr)
+		}
+		time.Sleep(time.Millisecond * 30) // тормозит выполнение, но зато прогресс бар красиво наполняется
+	}
+
+	printProgressBar(bytesToCopy, bytesToCopy, 40)
```

Comment:
```suggestion
        const defaultWidth = 40
	printProgressBar(bytesToCopy, bytesToCopy, defaultWidth)
```
Magic numbers сбивают с толку, такие переменные я бы рекомендовал переносить в блок констант и называть как то чтобы можно было понять что это и зачем